### ISSUE TITLE
[feature] expose raw audio and video data

### ISSUE THREAD

**Issue by davidzhao** [livekit/client-sdk-flutter#201]
**Is your feature request related to a problem? Please describe.**

Users would like to record audio and video locally in addition to publishing to the server. Currently audio and video data are not being exposed locally.

**Describe the solution you'd like**

When bytes are generated, we should expose a callback to receive the streams.

**Comment by Parfyonator** [comment]
It would be a very cool and useful feature. Really looking forward to use it.

BTW, @cloudwebrtc  are you planning to expose methods for tracks from the `flutter-webrtc` dependency or do something different? Currently `captureFrame` method of `MediaStreamTack` and `MediaStreamRecorder` (for video recording) are producing desaturated images and videos with artefacts on Android platform due to (maybe?) incorrect implementation of some native methods in WebRTC dependency. More details in this [issue](https://github.com/flutter-webrtc/flutter-webrtc/issues/496).

**Comment by cloudwebrtc** [comment]
@Parfyonator Here's a captureFrame method fix related to color bugs. maybe you can try if fix your issue
https://github.com/flutter-webrtc/flutter-webrtc/pull/1258

**Comment by shalom-aviv** [comment]
Good feature
Would be nice if it were possible to set different parameters for streaming and for the saved video.
For example
locally save a video track in 4k but stream it in the lowest quality.

**Comment by jtkeyva** [comment]
Great feature. Would also be nice to be able to use MLKit etc by adding access to the image stream from Dart.

**Comment by etegan** [comment]
Is this something we could expect in the near future?

**Comment by cloudwebrtc** [comment]
I should try to implement them in the next few months, which is related to native layer audio processing.

**Comment by jtkeyva** [comment]
yeah would be very useful to store the video or portions of it when you hit record locally as well

**Comment by EvertonMJunior** [comment]
Are there any updates on this?

**Comment by cloudwebrtc** [comment]
Some updates:
With this PR, the following issues will be resolved.

- You can register an Audio frame processor for third-party noise reduction, voice visualization,
- Register a Video Frame processor to process the Camera frame, add background blur, or other CV processing.

- register a sink for the audio/video track to receive frames for recording or pip display

https://github.com/flutter-webrtc/flutter-webrtc/pull/1700

**Comment by willsmanley** [comment]
that's amazing! thank you so much for this PR

tangential: next step for me is to consume raw audio bytes to make a visualizer widget, similar to Web AudioContext API. maybe need to use dart_fft but haven't looked into it yet. if anyone comes across this message and has any tips, please share!

**Comment by jezell** [comment]
@cloudwebrtc Awesome!!!

**Comment by jtkeyva** [comment]
@willsmanley 

try so loud and even better they are gonna support PCM soon

https://github.com/alnitak/flutter_soloud/issues/152#issuecomment-2480934191

**Comment by willsmanley** [comment]
> @willsmanley try so loud and even better they are gonna support PCM soon

perfect, this is why i love open source 😭

**Comment by jtkeyva** [comment]
> > @willsmanley try so loud and even better they are gonna support PCM soon
> 
> perfect! this is why i love open source 😭

@willsmanley please do share here or r/flutterdev when you get your visualizer widget working...sounds cool :)

**Comment by JinkeJ** [comment]
@cloudwebrtc how about generating custom frame without opening a camera? Like playing a file but with watermark.

**Comment by BrentCarter** [comment]
I'm looking to record each participant's audio, video, and screenshare tracks locally on each participant's broadcasting device so I can have recordings of the best possible quality, even if the streams used for the meeting in real time are lower quality or choppy due to network issues. I want to record at a resolution of my choice including full resolution e.g. up to 4K 60p on a phone even if the video resolution / framerate that gets streamed for the live meeting is lower. I then want to upload the locally recorded streams to cloud storage in the background.

@davidzhao posted this issue with title "expose raw audio and video data" and the post by @cloudwebrtc on Nov 7 seems to allow arbitrary processing of audio and video data but yet this issue has not been marked closed.

Is it possible for @davidzhao or @cloudwebrtc to clarify what remains to be done on this issue and if my scenario is possible with the completed features mentioned by @cloudwebrtc above? And if not, will it be possible when whatever remains to be done for this issue is completed?

**Comment by cloudwebrtc** [comment]
Hey @BrentCarter, Technically, this issue has been resolved. You can refer to this filter plugin and create a local recording plugin based on it. You will need to process these audio-video frames at the native level and record them as a video file.

https://github.com/livekit-examples/flutter-filter-plugin-example

**Comment by yashdeepsingh-create** [comment]
@cloudwebrtc The provided example is outdated as "class LivekitAudioFilter" requires "MediaStreamTrack? get processedTrack" when implements TrackProcessor, so can you help me here?

https://github.com/livekit-examples/flutter-filter-plugin-example/blob/dc555a8df3891769748410dcb4d2ef70fc7ea707/lib/src/audio_filter.dart#L5
